### ISSUE TITLE
GraphQLRequest type: add optional headers field

### ISSUE THREAD

**Issue by naseemkullah** [apollographql/apollo-client#10781]
### Issue Description

It appears that assigning `headers` to a `GraphQLRequest` operation yields +ve results. I suggest adding it to the type. as an option Record`<string, string>`

### Link to Reproduction

nope

**Comment by jerelmiller** [comment]
Hey @naseemkullah 👋 

Would you be able to add some code samples, a reproduction, or more detail about what you're looking for? I'm not sure I have enough information to understand your ask here.

**Comment by naseemkullah** [comment]
Hi @jerelmiller 👋 


Sure, I tried all sorts of things and only the following worked:

```ts
// Create an Apollo Link that intercepts requests
const baggageLink = setContext(
  (operation: GraphQLRequest & { headers?: Record<string, string> }) => {
    const user = getUser();
    
    // just some manipulation to get an obj with `{key: {value: value} }` structure
    if (user) {
      const baggageEntries = Object.entries({
        'user.name': user?.name,
        'user.email': user?.email,
      }).reduce((acc: Record<string, BaggageEntry>, [key, value]) => {
        if (value !== undefined && value !== null) {
          acc[key] = { value: value.toString() };
        }
        return acc;
      }, {});

      const baggage = propagation.createBaggage(baggageEntries);

      operation.headers = operation.headers ?? {};

      const propagatedContext = propagation.setBaggage(ROOT_CONTEXT, baggage);
      const propagator = new W3CBaggagePropagator();
      propagator.inject(propagatedContext, operation.headers, defaultTextMapSetter);
    }

    return operation;
  }
);
```

tl;dr I am exploring https://opentelemetry.io/docs/concepts/signals/baggage/ which requires a "carrier" for propagation, in this case it would be the request headers, only by passing it the operation.headers did this work

**Comment by jerelmiller** [comment]
@naseemkullah thanks! That's useful.

It looks like you're getting `headers` from the wrong place. Assuming you're using `HttpLink`, this is pulled from [`context`](https://github.com/apollographql/apollo-client/blob/499701310fd244877f35067c35ad0eae5dc7be70/src/link/http/createHttpLink.ts#L81), not the `operation` itself. Rather than using `operation.headers`, you'll want to `operation.context.headers` to get any headers sent from previous links or from a query. As a shortcut, you can get access to the previous context from the second argument to `setContext`

```ts
const baggageLink = setContext((operation, context) => {
  const headers = context.headers; // same as operation.context.headers
});
```

Something else to note is that the `setContext` link is expected to return a _context_ object, not the request operation, so its likely your `context` object is messed up somewhere down the line.

I'm familiar with OpenTelemetry, but I haven't used the libraries, so I'm not familiar with the baggage APIs. Does this modify the headers in some way? If so, you can return a `headers` property from your `setContext` call to ensure this is propagated (see the example in the [docs](https://www.apollographql.com/docs/react/api/link/apollo-link-context)):

```ts
const baggageLink = setContext((operation, context) => {
  const { headers } = context;

  // do something with headers

  return { headers: modifiedHeaders }
});
```

If you need to modify the `operation` itself and not just headers, I'd suggest instead using a plain `ApolloLink` to handle this. This might look something like this:

```ts
const baggageLink = new ApolloLink((operation, forward) => {
  const { headers } = operation.getContext();

  const modifiedOperation = // ...

  return forward(modifiedOperation);
});
```

Does this get you what you're looking for?

**Comment by naseemkullah** [comment]
> @naseemkullah thanks! That's useful.
> 
> It looks like you're getting `headers` from the wrong place. Assuming you're using `HttpLink`, this is pulled from [`context`](https://github.com/apollographql/apollo-client/blob/499701310fd244877f35067c35ad0eae5dc7be70/src/link/http/createHttpLink.ts#L81), not the `operation` itself. Rather than using `operation.headers`, you'll want to `operation.context.headers` to get any headers sent from previous links or from a query. As a shortcut, you can get access to the previous context from the second argument to `setContext`
> 
> ```ts
> const baggageLink = setContext((operation, context) => {
>   const headers = context.headers; // same as operation.context.headers
> });
> ```
> 
> Something else to note is that the `setContext` link is expected to return a _context_ object, not the request operation, so its likely your `context` object is messed up somewhere down the line (you can see we pass the result of your `setter` function to [`operation.setContext`](https://github.com/apollographql/apollo-client/blob/499701310fd244877f35067c35ad0eae5dc7be70/src/link/context/index.ts#L19) in this link).
> 
> I'm familiar with OpenTelemetry, but I haven't used the libraries, so I'm not familiar with the baggage APIs. Does this modify the headers in some way? If so, you can return a `headers` property from your `setContext` call to ensure this is propagated (see the example in the [docs](https://www.apollographql.com/docs/react/api/link/apollo-link-context)):
> 
> ```ts
> const baggageLink = setContext((operation, context) => {
>   const { headers } = context;
> 
>   // do something with headers
> 
>   return { headers: modifiedHeaders }
> });
> ```
> 
> If you need to modify the `operation` itself and not just headers, I'd suggest instead using a plain [`ApolloLink`](https://www.apollographql.com/docs/react/api/link/introduction) to handle this. This might look something like this:
> 
> ```ts
> const baggageLink = new ApolloLink((operation, forward) => {
>   const { headers } = operation.getContext();
> 
>   const modifiedOperation = // ...
> 
>   return forward(modifiedOperation);
> });
> ```
> 
> Does this get you what you're looking for?

It sure does, thanks @jerelmiller. I thought I had tried this approach already but upon trying harder it worked:

specifically this one:

```ts
const baggageLink = setContext((operation, context) => {
  const { headers } = context;

  // do something with headers

  return { headers: modifiedHeaders }
});
```

**Comment by naseemkullah** [comment]
worth noting, due to the nature of the [opentelemetry function](https://github.com/open-telemetry/opentelemetry-js/blob/053acb604c9d2b735bb267ff47724bb3af1ffd62/api/src/propagation/TextMapPropagator.ts#L129-L137) that modifies the "carrier" (req headers in this case), I have to set headers to an empty obj if its null:

```ts
    ctx.headers = ctx.headers ?? {};
```

**Comment by jerelmiller** [comment]
@naseemkullah awesome! Glad you were able to get it working! 

I'll have to look more at that OpenTelemetry stuff. I used to work at New Relic and saw a lot of that, but never directly used it myself. Cool stuff!

**Comment by naseemkullah** [comment]
> @naseemkullah awesome! Glad you were able to get it working!
>

Thanks again for the help, appreciate it!
 
> I'll have to look more at that OpenTelemetry stuff. I used to work at New Relic and saw a lot of that, but never directly used it myself. Cool stuff!

oh nice!

**Comment by naseemkullah** [comment]
Hey @jerelmiller i just realized i may want to specify to which uris the baggage headers should be added to (i may want to avoid adding this when making reqs to 3rd parties) ...is there a way to get the `uri` within setContext()?

**Comment by phryneas** [comment]
You can set a new uri using context, but not get the "default" uri, as that is only an option to your HttpLink, which is further down the chain.

What you *can* do instead of using context would be to pass a function into `uri` when creating a `HttpLink`.

Something along these lines:

```js
const httpLink = new HttpLink({
  uri: (operation) => {
    const { operationName, variables } = operation;
    return `${process.env.REACT_APP_API_URL}/graphql?operationName=${operationName}&variables=${JSON.stringify(
      variables
    )}`;
  }
});
```

**Comment by phryneas** [comment]
Thinking further, you could also try wrapping the `fetch` version in a custom implementation and setting your headers there instead. At that point, you don't have access to the context anymore, so you would probably have to use that `uri` callback function to add information to the url, and then in a `fetch` wrapper function remove that information again and put it into headers instead 🤔 

So you'd end up with a hack like this:
```ts
const link = new HttpLink({
  uri: (operation) => {
    const { operationName } = operation;
    return `${process.env.REACT_APP_API_URL}/graphql?operationName=${operationName}`;
  },
  fetch(input, init) {
    // add extra headers based on `input`
    if (typeof input !== 'string') { return fetch(input, init) }
    const url = new URL(input)
    const headers = new Headers(init?.headers)
    const operationName = url.searchParams.get('operationName')
    url.searchParams.delete('operationName')
    headers.append('extra-header', operationName)
    return fetch(url.toString(), { ...init, headers })
  },
})
```
